pkg/controllers/scheduling/scheduling_controller.go

```diff
@@ -0,0 +1,331 @@
+			continue
+		}
+
+		selectedClusters.Insert(decision.ClusterName)
+	}
+
+	// no work if all decisions have already been scheduled
+	queueKey := fmt.Sprintf("%s/%s", placementDecision.Namespace, placementDecision.Name)
+	numOfDecisions := len(placementDecision.Status.Decisions)
+	if numOfDecisions == selectedClusters.Len() {
+		c.rateLimiter.Forget(queueKey)
+		return nil
+	}
+
+	// get the owner placement
+	placementName := placementDecision.Labels[placementLabel]
+	if len(placementName) == 0 {
+		return fmt.Errorf("no placement label on placementdecision %q", queueKey)
+	}
+	placement, err := c.placementLister.Placements(placementDecision.Namespace).Get(placementName)
+	if err != nil {
+		return err
+	}
+
+	// get all feasible clusters for the placement
+	clusters, err := getFeasibleClusters(placement, c.clusterLister, c.clusterSetLister, c.clusterSetBindingLister)
+	if err != nil {
+		return err
+	}
+
+	// make decisions for the unscheduled decisions
+	newPlacementDecision := placementDecision.DeepCopy()
+	for index, decision := range newPlacementDecision.Status.Decisions {
+		if len(decision.ClusterName) != 0 {
+			continue
+		}
+
+		decision := scheduleOne(clusters, selectedClusters)
+		// stop if no more cluster is available
+		if decision == nil {
+			break
+		}
+
+		newPlacementDecision.Status.Decisions[index] = *decision
+		selectedClusters.Insert(decision.ClusterName)
+	}
+
+	// update the placementdecision if necessary
+	if reflect.DeepEqual(placementDecision.Status.Decisions, newPlacementDecision.Status.Decisions) {
+		c.requeueOrForget(queueKey, syncCtx.Queue(), numOfDecisions, selectedClusters.Len())
+		return nil
+	}
+
+	if _, err := c.clusterClient.ClusterV1alpha1().PlacementDecisions(newPlacementDecision.Namespace).
+		UpdateStatus(ctx, newPlacementDecision, metav1.UpdateOptions{}); err != nil {
+		return err
+	}
+
+	// update the status of the placement if necessary
+	if err := updatePlacementStatus(ctx, placement, numOfDecisions, selectedClusters.Len(), c.clusterClient); err != nil {
+		return err
+	}
+
+	c.requeueOrForget(queueKey, syncCtx.Queue(), numOfDecisions, selectedClusters.Len())
+	return nil
+}
+
+// requeueOrForget requeues or forgets the given queue key according to the values of numOfDecisions
+// and numOfSelectedClusters
+func (c *decisionSchedulingController) requeueOrForget(queueKey string, queue workqueue.RateLimitingInterface, numOfDecisions, numOfSelectedClusters int) {
+	// requeue the placementdecison if there exists any unscheduled decision
+	if numOfDecisions > numOfSelectedClusters {
+		queue.AddAfter(queueKey, c.rateLimiter.When(queueKey))
+		return
+	}
+	c.rateLimiter.Forget(queueKey)
+}
+
+// scheduleOne picks up one cluster from given clusters. A cluster can not be picked up if it has alreay been selected.
+// Return nil if there is no more cluster available.
+func scheduleOne(clusters []*clusterapiv1.ManagedCluster, selectedClusters sets.String) *clusterapiv1alpha1.ClusterDecision {
```

Comment:
It might be very expensive to filter clusters in each call of scheduleOne(). Why not we just do filtering outside of scheduleOne() and then pass the the feasible clusters as parameter when calling scheduleOne()?

Reply:
Code updated as suggested.